titus-server-master/src/main/java/com/netflix/titus/master/kubernetes/controller/PodGcController.java

```diff
@@ -0,0 +1,246 @@
+        for (V1Pod pod : podsPastDeletionTimestampToGc) {
+            gcPod(pod);
+        }
+        logger.info("Finished pods past deletion timestamp GC");
+    }
+
+    @VisibleForTesting
+    boolean isPodPastDeletionTimestamp(V1Pod pod) {
+        V1PodSpec spec = pod.getSpec();
+        V1ObjectMeta metadata = pod.getMetadata();
+        if (spec == null || metadata == null) {
+            return false;
+        }
+
+        Long terminationGracePeriodSeconds = spec.getTerminationGracePeriodSeconds();
+        DateTime deletionTimestamp = metadata.getDeletionTimestamp();
+        if (terminationGracePeriodSeconds == null || deletionTimestamp == null) {
+            return false;
+        }
+
+        long terminationGracePeriodMs = terminationGracePeriodSeconds * 1000L;
+        return clock.isPast(deletionTimestamp.getMillis() + terminationGracePeriodMs
+                + configuration.getPodsPastTerminationGracePeriodMs());
+    }
+
+    /**
+     * GC pods in Pending phase with a deletion timestamp.
+     */
+    private void gcPendingPodsWithDeletionTimestamp(List<V1Pod> pods) {
+        List<V1Pod> pendingPodsWithDeletionTimestampToGc = pods.stream()
+                .filter(this::isPodInPendingPhaseWithDeletionTimestamp)
+                .collect(Collectors.toList());
+
+        logger.info("Attempting to GC {} pending pods: {} with deletion timestamp", pendingPodsWithDeletionTimestampToGc.size(),
+                pendingPodsWithDeletionTimestampToGc);
+        pendingPodsWithDeletionTimestampToGcGauge.set(pendingPodsWithDeletionTimestampToGc.size());
+        for (V1Pod pod : pendingPodsWithDeletionTimestampToGc) {
+            gcPod(pod);
+        }
+        logger.info("Finished pending pods with deletion timestamp GC");
+    }
+
+    @VisibleForTesting
+    boolean isPodInPendingPhaseWithDeletionTimestamp(V1Pod pod) {
+        if (pod == null || pod.getMetadata() == null || pod.getStatus() == null) {
+            return false;
+        }
+        DateTime deletionTimestamp = pod.getMetadata().getDeletionTimestamp();
+        return deletionTimestamp != null && PENDING.equalsIgnoreCase(pod.getStatus().getPhase());
+    }
+
+    private void gcPod(V1Pod pod) {
+        if (pod != null) {
+            gcPod(KubeUtil.getMetadataName(pod.getMetadata()));
+        }
+    }
+
+    private void gcPod(String podName) {
+        if (StringExt.isNotEmpty(podName)) {
+            try {
+                kubeApiFacade.getCoreV1Api().deleteNamespacedPod(
+                        podName,
+                        DEFAULT_NAMESPACE,
+                        null,
+                        null,
+                        0,
+                        null,
+                        BACKGROUND,
+                        null
+                );
+            } catch (JsonSyntaxException e) {
+                // this is probably successful. the generated client has the wrong response type
```

Comment:
Do we know how frequently and when this case could happen? Does it make sense to log at DEBUG level in case we need to troubleshoot a GC issue?

Reply:
it happens 100% of the time so I don't think we need a log line